feat: add log format configuration options through helm chart

mowies opened the pull request:
## This PR
- adds a new `sidecarConfiguration.logFormat` helm value that is handed over to the operator as an env var
- sets the default sidecar log format to json both in code and in the CRD itself, as well as the helm chart

Fixes #334

james-milligan commented on apis/core/v1alpha1/flagsourceconfiguration_types.go:
```diff
@@ -92,6 +94,11 @@ type FlagSourceConfigurationSpec struct {
+	// DefaultLogFormat allows for the sidecar log format to be overridden, defaults to 'json'
+	// +optional
+	// +kubebuilder:default:=json
+	DefaultLogFormat string `json:"logFormat"`
```
this change also should be introduced into the `v1alpha2` version, with the appropriate changes made in the `flagsourceconfiguration_conversion.go` file.

mowies replied:
done

james-milligan commented:
The `Merge` method in `apis/core/v1alpha1/flagsourceconfiguration_types.go` should be extended to ensure this configuration isn't dropped.

```
if new.DefaultLogFormat != "" {
	fc.DefaultLogFormat = new.DefaultLogFormat
}
```

I may also be worthwhile adding this configuration to one of the existing tests in `/webhooks/pod_webhook_test.go` as this will catch the above issue in future PRs

james-milligan commented on apis/core/v1alpha1/flagsourceconfiguration_types.go:
```diff
@@ -92,6 +94,11 @@ type FlagSourceConfigurationSpec struct {
+	// DefaultLogFormat allows for the sidecar log format to be overridden, defaults to 'json'
+	// +optional
+	// +kubebuilder:default:=json
+	DefaultLogFormat string `json:"logFormat"`
```
```suggestion
	LogFormat string `json:"logFormat"`
```

mowies replied:
done

mowies commented:
@james-milligan thanks for the review!
I fixed your remarks, adjusted the mutating webhook tests, and also re-generated the helm chart (not sure if that's needed or not)
pls re-review :)

beeme1mr commented on apis/core/v1alpha1/flagsourceconfiguration_types.go:
```diff
@@ -38,15 +38,17 @@ const (
 	defaultImage                     string = "ghcr.io/open-feature/flagd"
-	// `INPUT_FLAGD_VERSION` is replaced in the `update-flagd` Makefile target
-	defaultTag             string           = "INPUT_FLAGD_VERSION"
+	// `v0.3.4` is replaced in the `update-flagd` Makefile target
+	defaultTag             string           = "v0.3.4"
```
```suggestion
	// `INPUT_FLAGD_VERSION` is replaced in the `update-flagd` Makefile target
	defaultTag             string           = "INPUT_FLAGD_VERSION"
```

This is automatically updated during the build process.

toddbaert replied:
Yes in fact it will fail the build (I believe we have some check around this).

james-milligan commented on apis/core/v1alpha2/flagsourceconfiguration_types.go:
```diff
@@ -59,6 +59,11 @@ type FlagSourceConfigurationSpec struct {
+
+	// LogFormat allows for the sidecar log format to be overridden, defaults to 'json'
+	// +optional
+	// +kubebuilder:default:=json
```
will this cause issues when we merge the CRs? currently we are checking if `new.LogFormat != ""`, if this always sets a default value of `"json"` we will override configurations with this default on each merge

mowies replied:
i can also remove it again. i saw that you don't use it for anything else

james-milligan replied:
my vote would be to remove it, but happy to update the merge function to check for non-default

mowies replied:
i removed it :)

mowies commented:
the failing test work for me locally... can somebody maybe start a re-run?

skyerus commented:
> the failing test work for me locally... can somebody maybe start a re-run?

Done! Note: there's an issue for this flakey test https://github.com/open-feature/open-feature-operator/issues/347

toddbaert reviewed (approved):
I didn't manually test this one but it looks good to me; tests are passing now and it looks like @james-milligan 's concerns (including the conversion) have been resolved.

james-milligan commented on chart/open-feature-operator/README.md:
```diff
@@ -43,16 +43,17 @@ The command removes all the Kubernetes components associated with the chart and
 | `sidecarConfiguration.providerArgs`      | `""`  | Used to append arguments to the sidecar startup command. This value is a comma separated string of key values separated by '=', e.g. `key=value,key2=value2` results in the appending of `--sync-provider-args key=value --sync-provider-args key2=value2` |
-| `sidecarConfiguration.defaultSyncProvider`      | `kubernetes`  | Sets the value of the `XXX_SYNC_PROVIDER` environment variable for the injected sidecar container. There are 3 valid sync providers: `kubernetes`, `filepath` and `http` |
+| `sidecarConfiguration.defaultSyncProvider`      | `kubernetes`  | Sets the value of the `XXX_SYNC_PROVIDER` environment variable for the injected sidecar container. There are 3 valid sync providers: `kubernetes`, `filepath` and `http`                                                                                  |
+| `sidecarConfiguration.logFormat` | `json` | Sets the log format used by the flagD sidecar. `json` and `console` are possible                                                                                                                                                                          |
```
```suggestion
| `sidecarConfiguration.logFormat` | `json` | Sets the value of the `XXX_LOG_FORMAT` environment variable for the injected sidecar container.                                                                                                                                                                          |
```

james-milligan replied:
Where possible its best to keep this documentation sidecar agnostic

skyerus reviewed (approved):
LGTM, thanks :)
